c++/src/kj/memory.h

```diff
@@ -656,6 +697,172 @@ class SpaceFor {
   };
 };
 
+// =======================================================================================
+// Pin<T>
+
+template <typename T>
+class Ptr;
+
+template <typename T>
+class Pin {
+// Pin<T> is a smart, in-place storage for T. 
+//
+// Pin<T> should be created on the stack or used as a data member. It should not be 
+// allocated on the heap.
```

Comment:
Should we enforce this by overriding operator new and operator new[]?

Reply:
There's an existing kj utility for enforcing this rule.... `kj::requireOnStack(...)` I think... not sure if it's appropriate for this case tho. @kentonv could say tho

Reply:
kj::requireOnStack wouldn't work, since `kj::Pin` could be techically on the heap as a data field (which is fine).

overriding operators is best here I think.